packages/lexical-selection/src/range-selection.ts

```diff
@@ -7,304 +7,81 @@
-        const targetElement = createElement();
-        targetElement.setFormat(parent.getFormatType());
-        targetElement.setIndent(parent.getIndent());
-        elements.push(targetElement);
-        elementMapping.set(parentKey, targetElement);
-        // Move node and its siblings to the new
-        // element.
-        parent.getChildren().forEach((child) => {
-          targetElement.append(child);
-          movedLeafNodes.add(child.getKey());
-        });
-        $removeParentEmptyElements(parent);
-      }
-    } else if (emptyElements.has(node.getKey())) {
-      const targetElement = createElement();
-      targetElement.setFormat(node.getFormatType());
-      targetElement.setIndent(node.getIndent());
-      elements.push(targetElement);
-      node.remove(true);
-    }
-  }
-
-  if (wrappingElement !== null) {
-    for (let i = 0; i < elements.length; i++) {
-      const element = elements[i];
-      wrappingElement.append(element);
-    }
-  }
-
-  // If our target is Root-like, let's see if we can re-adjust
-  // so that the target is the first child instead.
-  if ($isRootOrShadowRoot(target)) {
-    if (targetIsPrevSibling) {
-      if (wrappingElement !== null) {
-        target.insertAfter(wrappingElement);
-      } else {
-        for (let i = elements.length - 1; i >= 0; i--) {
-          const element = elements[i];
-          target.insertAfter(element);
-        }
-      }
-    } else {
-      const firstChild = target.getFirstChild();
-
-      if ($isElementNode(firstChild)) {
-        target = firstChild;
-      }
-
-      if (firstChild === null) {
-        if (wrappingElement) {
-          target.append(wrappingElement);
-        } else {
-          for (let i = 0; i < elements.length; i++) {
-            const element = elements[i];
-            target.append(element);
-          }
-        }
-      } else {
-        if (wrappingElement !== null) {
-          firstChild.insertBefore(wrappingElement);
-        } else {
-          for (let i = 0; i < elements.length; i++) {
-            const element = elements[i];
-            firstChild.insertBefore(element);
-          }
-        }
-      }
-    }
-  } else {
-    if (wrappingElement) {
-      target.insertAfter(wrappingElement);
-    } else {
-      for (let i = elements.length - 1; i >= 0; i--) {
-        const element = elements[i];
-        target.insertAfter(element);
-      }
-    }
-  }
-
-  const prevSelection = $getPreviousSelection();
-
-  if (
-    $isRangeSelection(prevSelection) &&
-    isPointAttached(prevSelection.anchor) &&
-    isPointAttached(prevSelection.focus)
-  ) {
-    $setSelection(prevSelection.clone());
-  } else {
-    selection.dirty = true;
-  }
+function isBlock(node: LexicalNode) {
+  return (
+    $isElementNode(node) &&
+    ($isParagraphNode(node) ||
+      $isHeadingNode(node) ||
+      $isListItemNode(node) ||
+      $isQuoteNode(node) ||
+      $isCodeNode(node))
+  );
```

Comment:
Is the purpose of this function to check if the node can be inline or not? Can't we simply use `isInline` method?

```suggestion
  return !node.isInline();
```

_If this works, we do not even need to define this as a separate function._

Reply:
_Without getting deep into it._ The reason I proposed to use `isInline` is that `isBlock` function made me think that the logic uses it to attempt to skip steps when something is inline. It might be better to rename it to something different if that's not the case.

Reply:
Actually, there are many types of nodes. Think of table, row, cell, list, listitem... etc.
One option would be to add a $isBlockNode function to each node, as was done with ShadowRoot. But I am reluctant to add new concepts and complexity to the definition of a node.
I think something pretty close to what I put in the comment above should work.